pl_env.h

```diff
@@ -0,0 +1,220 @@
+ *      variable is missing. This is mitigated by improving the error message
+ *      for the most common error, failed playback due to a nonexistent path
+ *      caused by a missing environment variable.
+ *   3. It is not possible to implement a mechanism for fallback paths within
+ *      cmus for searching for individual tracks.
+ *
+ * - What are some potential uses of pl_env_vars?
+ *   1. Syncing a $CMUS_HOME with multiple devices with different home folders.
+ *   2. Syncing a $CMUS_HOME with multiple devices with one or more different
+ *      music paths.
+ *   3. The above case, but also with a path which only exists on certain
+ *      devices.
+ *   4. Easily relocating the cache/library/playlists while preserving metadata
+ *      including the play count. This can be done by setting pl_env_vars,
+ *      exiting cmus, updating the env var to the new path, starting cmus,
+ *      clearing pl_env_vars, and exiting cmus again.
+ */
+
+/**
+ * PL_ENV_DELIMITER surrounds env var substitutions at the beginning of paths.
+ */
+#define PL_ENV_DELIMITER '\x1F'
+
+/**
+ * pl_env_save checks the base path against the configured environment
+ * variables, replaces the first match with a substitution, and returns a
+ * malloc'd copy of the result. If there isn't any valid match or the path
+ * already contains a substitution, a copy of the original path is returned
+ * as-is.
+ */
+char *pl_env_save(const char *path);
+
+/**
+ * pl_env_restore returns a malloc'd copy of path, with the environment variable
+ * substitution. The provided path must use forward slashes, and begin with a
+ * slash or a substitution followed by a slash (which will always be true for
+ * library paths within cmus). If the path does not have a substitution, the
+ * original path is returned. If the environment variable does not exist or is
+ * invalid, the original path is also returned.
+ */
+char *pl_env_restore(const char *path);
+
+/**
+ * pl_env_var returns a pointer to the start of the substituted environment
+ * variable name, or NULL if it is not present. If a variable is present and
+ * out_length is not NULL, it is set to the length of the variable name. The
+ * remainder of the path will be at pl_env_var_remainder(path, out_length).
+ */
+const char *pl_env_var(const char *path, int *out_length);
+
+/**
+ * pl_env_var_remainder returns a pointer to the remainder of the path after the
+ * substitution. See pl_env_var.
+ */
+const char *pl_env_var_remainder(const char *path, int length);
+
+/**
+ * pl_env_var_str is like pl_env_var, but it returns a malloc'd null-terminated
+ * string instead. If remainder is not NULL, it is set to point to the remainder
+ * of the path.
+ */
+char *pl_env_var_str(const char *path, const char **remainder);
+
+/**
+ * pl_env_var_len returns the length of the substituted environment variable
+ * name, if present. Otherwise, it returns 0.
+ */
+int pl_env_var_len(const char *path);
+
+#endif
```

Comment:
Should I switch to static thread-local buffers to store the results? It's safe for how these functions are currently used (only one track is ever processed at a time), and it would essentially eliminate the overhead of this option if it's enabled.

Reply:
I've decided to leave it as-is since the performance benefit of using buffers is negligible, and it's safer not to use them.